test/es-module/test-esm-loader-resolve-type.mjs

```diff
@@ -1,44 +1,44 @@
-// Flags: --loader ./test/fixtures/es-module-loaders/hook-resolve-type.mjs
-import { allowGlobals } from '../common/index.mjs';
+import { spawnPromisified } from '../common/index.mjs';
+import * as tmpdir from '../common/tmpdir.js';
 import * as fixtures from '../common/fixtures.mjs';
-import { strict as assert } from 'assert';
-import * as fs from 'fs';
-
-allowGlobals(global.getModuleTypeStats);
-
-const { importedESM: importedESMBefore,
-        importedCJS: importedCJSBefore } = await global.getModuleTypeStats();
-
-const basePath =
-  new URL('./node_modules/', import.meta.url);
-
-const rel = (file) => new URL(file, basePath);
-const createDir = (path) => {
-  if (!fs.existsSync(path)) {
-    fs.mkdirSync(path);
-  }
-};
+import { deepStrictEqual } from 'node:assert';
+import { mkdir, rm, cp } from 'node:fs/promises';
+import { execPath } from 'node:process';
 
+const base = tmpdir.fileURL(`test-esm-loader-resolve-type-${(Math.random() * Date.now()).toFixed(0)}`);
 const moduleName = 'module-counter-by-type';
-const moduleDir = rel(`${moduleName}`);
+const moduleURL = new URL(`${base}/node_modules/${moduleName}`);
 try {
-  createDir(basePath);
-  createDir(moduleDir);
-  fs.cpSync(
-    fixtures.path('es-modules', moduleName),
-    moduleDir,
+  await mkdir(moduleURL, { recursive: true });
+  await cp(
+    fixtures.path('es-modules', 'module-counter-by-type'),
+    moduleURL,
     { recursive: true }
   );
 
-
-  await import(`${moduleName}`);
+  deepStrictEqual(await spawnPromisified(
+    execPath,
+    [
+      '--no-warnings',
+      '--input-type=module',
+      '--eval',
+      `import { getModuleTypeStats } from ${JSON.stringify(fixtures.fileURL('es-module-loaders', 'hook-resolve-type.mjs'))};
+      const before = getModuleTypeStats();
+      await import(${JSON.stringify(moduleName)});
+      const after = getModuleTypeStats();
+      console.log(JSON.stringify({ before, after }));`,
+    ],
+    { cwd: base },
+  ), {
+    stderr: '',
+    stdout: JSON.stringify({
+      before: { importedESM: 0, importedCJS: 0 },
+      // Dynamic import in the eval script should increment ESM counter but not CJS counter
+      after: { importedESM: 1, importedCJS: 0 },
+    }) + '\n',
+    code: 0,
+    signal: null,
+  });
```

Comment:
Wrapping this in `deepStrictEqual` looks very strange and is difficult to read—the expectation argument looks like an argument to `spawnPromisified`). If you want to do the deepStrictEqual, I think readability would be much improved by first capturing the output of `spawnPromisified` and then comparing it:

```js
const output = await spawnPromisified(…);

deepStrictEqual(output, {
  …
});
```

Reply:
This was an @aduh95 refactor 😄

Reply:
I think converting it to `spawnPromisified` is a good idea—I had done when you asked me to finish this up. Just the deepStrictEqual as a wrapper is my concern.

Reply:
I tried changing it to compare against `output` but I get `AssertionError [ERR_ASSERTION]: Values have same structure but are not reference-equal`. I think this is good enough for now.